protofsm/daemon_events.go

```diff
@@ -0,0 +1,79 @@
+package protofsm
+
+import (
+	"github.com/btcsuite/btcd/btcec/v2"
+	"github.com/btcsuite/btcd/wire"
+	"github.com/lightningnetwork/lnd/fn"
+	"github.com/lightningnetwork/lnd/lnwire"
+)
+
+// DaemonEvent is a special event that can be emmitted by a state transition
+// function. A state machine can use this to perform side effects, such as
+// sending a message to a peer, or broadcasting a transaction.
+type DaemonEvent interface {
+	daemonSealed()
+}
+
+// DaemonEventSet is a set of daemon events that can be emitted by a state
+// transition.
+type DaemonEventSet []DaemonEvent
+
+// DaemonEvents is a special type constraint that enumerates all the possible
+// types of daemon events.
+type DaemonEvents interface {
+	SendMsgEvent[any] | DisableChannelEvent | BroadcastTxn
+}
+
+// SendPredicate is a function that returns true if the target message should
+// sent.
+type SendPredicate = func() bool
```

Comment:
Sure, I guess could call like a `BoolCallback` or something? Or just a `ContinuationFunc`? 

I think this is useful for areas where we don't have the new hook concept. Eg: something waiting for a channel to be added to the graph before it acts. 

In the RBF closer, I use this to hook into the case where no dangling updates exist (can send `shutdown): https://github.com/lightningnetwork/lnd/blob/43386d5643f961f948bf95513933c7d5a72fc74e/peer/chan_observer.go#L46-L51

Alternatively, we can use the hooks to send an event into the state machine once the state has been achieved. Then we have a new transition to just handle that event (send the shutdown). I slightly prefer it as is though, as the code reads as more imperative. Otherwise, the state machine would need more knowledge of hooks, and the ability of the hook to do things limit emit a daemon event, wherein the only way to emit that today is as a return value (the state transitions don't have handle on the thing executing it, they're effectively sandboxed).